src/metabase/models/field_values.clj

```diff
@@ -150,20 +156,39 @@
-        field-name   (or (:name field) (:id field))]
+  (classloader/require 'metabase.db.metadata-queries)
+  (let [field-values          (FieldValues :field_id (u/the-id field))
+        field-name            (or (:name field) (:id field))
+        full-values           (distinct-values field false)
+        values-exceed-length? (values-exceed-total-max-length? full-values)
+        values                (when-not values-exceed-length?
+                                full-values)]
+    ;; exceeded_limit=true for a field with `has_field_values=:list` means that the list of values we cached in [[FieldValues]]
+    ;; for that field are is all possible values. This is used by UI to decide which field it needs to call the search API
+    ;; when user find values.
+    ;; exceeded_limit=true when:
+    ;; 1. the number of distinct values of this field exceeded [[metabase.db.metadata-queries/absolute-max-distinct-values-limit]]
+    ;;    -> thus the list we stored in [[FieldValues]] is just a subset of all possible values
+    ;; 2. the total legnth of all values exceeded [[total-max-length]]
+    ;;    -> in this case we don't store FieldValues at all
+    (when (and (= (:has_field_values field) :list)
+               (or values-exceed-length?
+                   (>= (count values)
+                      @(resolve 'metabase.db.metadata-queries/absolute-max-distinct-values-limit))))
+      (db/update! 'Field (u/the-id field) :exceeded_limit true))
     (cond
-      ;; If this Field is marked `auto-list`, and the number of values in now over the list threshold, we need to
-      ;; unmark it as `auto-list`. Switch it to `has_field_values` = `nil` and delete the FieldValues; this will
-      ;; result in it getting a Search Widget in the UI when `has_field_values` is automatically inferred by the
-      ;; [[metabase.models.field/infer-has-field-values]] hydration function (see that namespace for more detailed
+      ;; If this Field is marked `auto-list`, and the number of values in now over the [[auto-list-cardinality-threshold]] or
+      ;; the accumulated length of all values exceeded the [[total-max-length]] threshold
+      ;; we need to unmark it as `auto-list`. Switch it to `has_field_values` = `nil` and delete the FieldValues;
+      ;; this will result in it getting a Search Widget in the UI when `has_field_values` is automatically inferred
+      ;; by the [[metabase.models.field/infer-has-field-values]] hydration function (see that namespace for more detailed
       ;; discussion)
       ;;
       ;; It would be nicer if we could do this in analysis where it gets marked `:auto-list` in the first place, but
       ;; Fingerprints don't get updated regularly enough that we could detect the sudden increase in cardinality in a
       ;; way that could make this work. Thus, we are stuck doing it here :(
-      (and (> (count values) auto-list-cardinality-threshold)
+      (and (or values-exceed-length?
```

Comment:
this is actually a case we didn't catch before,
previously if an `auto-list` field suddenly has the total length of all distinct values more than 10000, we will skip this step.

the correct way to do it is we need to remove the field values and change `has_field_values` to `nil` so that UI will display a search widget for it.

-- I also added this case in the test